### PR TITLE
Improved SweetHome 3D importer

### PR DESCRIPTION
**Pull request by rostskadat**
# Problem

The current SweetHome 3D file importer does not import all the elements such as `pieceOfFurniture`, `light`, `camera`, etc.

# Solution

This pull request extend the existing importer in order to allow importing of some of the missing elements. It also expose a setting page under the `Import/Export`. It finally integrate with the `Render` (if available) in order to bootstrap a `Povray` project from the imported elements.

### CODE REVIEW

**Comment by yorikvanhavre** [comment]
This is great and highly welcome!!

Due to the size of this PR and the need for testing, though, I think it's best to schedule it for after the imminent 1.0 release...

**Comment by adrianinsaval** [comment]
yeah we need to upgrade that, github has probably moved ubuntu-latest to 24.04. I'll make a PR shortly

**Comment by adrianinsaval** [comment]
fixed CI at 9dd105b16a4c3317a90608f994906ad29d630331 so I rebased this PR

File: src/Mod/BIM/importers/importSH3DHelper.py
Comment (luzpaz) [inline src/Mod/BIM/importers/importSH3DHelper.py:911]: typo: should be 'between'

File: src/Mod/BIM/importers/importSH3DHelper.py
Comment (luzpaz) [inline src/Mod/BIM/importers/importSH3DHelper.py:1060]: typo: 'containing'

File: src/Mod/BIM/importers/importSH3DHelper.py
Comment (luzpaz) [inline src/Mod/BIM/importers/importSH3DHelper.py:1250]: in SweetHome3d '*s*' ? Is this a typo?

**Comment by luzpaz** [comment]
All typos found by CI:   
```shell
Warning: src/Mod/BIM/importers/importSH3DHelper.py:794: betwen ==> between
Warning: src/Mod/BIM/importers/importSH3DHelper.py:911: betwen ==> between
Warning: src/Mod/BIM/importers/importSH3DHelper.py:1060: containg ==> containing
Warning: src/Mod/BIM/importers/importSH3DHelper.py:1278: bouding ==> bounding
Warning: src/Mod/BIM/importers/importSH3DHelper.py:1291: correspondance ==> correspondence
```

**Comment by rostskadat** [comment]
@luzpaz  I'm having trouble writing a test for this functionality.
On my local machine I can run the newly created test with 
```shell
/mnt/build/bin/FreeCAD -t TestArch.ArchTest.testImportSH3D
```
And it run properly

However it fails the build pipeline with the message:

```
> from importers.importSH3DHelper import SH3DImporter
Module not found importers.importSH3DHelper
```

It seems that the `sys.path` is different on a local build and the build pipeline.
Could you point me toward a test that import an external file (I looked at Draft's  import SVG, but it's actually an empty test)

Thanks for the information.

**Comment by luzpaz** [comment]
@chennes :point_up_2:

**Comment by rostskadat** [comment]
Hello @yorikvanhavre ,

Great, I'll therefore add the test back into TestArch.

Thanks for your help

**Comment by luzpaz** [comment]
Great effort here @rostskadat !

**Comment by yorikvanhavre** [comment]
> Great, I'll therefore add the test back into TestArch.

Writing tests is always a good idea :)

**Comment by luzpaz** [comment]
Any volunteers to document this feature ?

**Comment by rostskadat** [comment]
> Any volunteers to document this feature ?

Hello @luzpaz ,

Could you expand on what you had in mind. I'm definitely interested in documenting that feature.
If you give me an example of a good feature documentation, I'll try my best to do the same for this one.

Thanks for your help

**Comment by yorikvanhavre** [comment]
> Could you orientate me toward a unit test with this characteristics (using an external file that is not really a source file)?

The way you did it seems good to me, it should work... `No such file or directory: '/usr/local/Mod/BIM/importers/samples/Sample.sh3d'` seems weird to me, though... I don't think any FreeCAD file should be installed in /usr/local/Mod. I suspect the problem here is not in your change in the BIM CMake file, but in the CMake setup of your build.

I would try this:

- Do a simple "make", not "make install". You can then launch FreeCAD with "./bin/FreeCAD". Check that your sample file is there in ./Mod/BIM/importers
- Change your cmake prefix to some user folder (ex. /home/YOUR_USER/FreeCAD) to test without "harming" your system

**Comment by rostskadat** [comment]
Hello @tritao, 

Thanks for your kind words. As you have noticed some elements (furniture such as door and windows) are not imported by default. I'll try to figure this out.

As for joining the walls I have encountered a certain number of corner cases that do not allow me to turn this on by default. The main obstacle is a behavior with Part::Sweep that I do not properly understand nor know how to work around (https://forum.freecad.org/viewtopic.php?t=92499). 

The other topic is documentation, where I'm not really sure what would the ideal "User Documentation".

Once these are solved, I think the importer would be in rather good shape to be considered for inclusion.

**Comment by rostskadat** [comment]
Hello @tritao, thanks for your insight. Modified the code to include your contribution.
However there is a more general issue between the Door&Windows in SH3D and in FreeCAD. 
Since in SH3D they are basically furniture, we will never have the 1-to-1 correspondence between a SH3D Door&Window and a FreeCAD one. 
I am actually skeptical of the "lookup the catalog id" strategy. If the goal of the import is to have the same aesthetic as in SH3D then we are probably doomed to fail. If the goal of the import is provide a "Door&Windows" placeholder in an architectural drawing, then we might be better of with one generic, IFC-compliant Door and one generic, IFC-compliant Window. 
What do you think? What kind of down processing are you aware of?

> Found another issue, the window/door handling did not recognize mirror walls, along with not setting the correct IFC attributes for doors. Here is an initial patch that fixes those issues.
> 
> ...
> 
> There are other improvements I think could be done, such as better checking of door types, along with recognizing double floor-to-human-height windows as double doors (Sweet Home doesn't seem to provide double door models).
> 
> Overall I wish this PR could be merged right away as the code is already high quality and would ease future collaboration on this.

**Comment by rostskadat** [comment]
Hello @tritao, yes the join wall flag is unreasonably tricky.

There are a couple of issues still to be ironed out to get it visually correct in all cases.
However, as for the "Door&Windows" import I'm confronted to a dilemma.
Allow me to share my doubts and ask your thoughts on it.

SH3D is probably mainly used for its visual appeal and drafting capabilities.
It does make sense for SH3D to join walls with a bezel, for the aesthetics  appeal.
FC is probably mainly used for its architectural drawing capabilities.
I doubt that any actual architectural drawing has any bezel-ed wall. Only straight corner.

Therefore I'm not sure whether we should strive to be visually faithful to SH3D when importing or if we should strive to be architecturally correct for downstream processes.

Maybe someone with an architectural background could shed light on the best path forward. What do you think?

> By the way, I am still seeing an issue with the corners, even with the Arch wall join option enabled, see the following image:
> 
> ...
> 
> [casa2.sh3d.zip](https://github.com/user-attachments/files/17952642/casa2.sh3d.zip)

**Comment by maxwxyz** [comment]
@rostskadat PR has a conflict

**Comment by maxwxyz** [comment]
@rostskadat can you resolve the conflict?

**Comment by yorikvanhavre** [comment]
I'm turning this to Draft status while there are conflicts and issues to solve, so it gets out of the merge queue. @rostskadat when you feel it's ready for merge, just hit the "ready for review" button!

**Comment by chennes** [comment]
@yorikvanhavre this is now ready for your re-review.

**Comment by yorikvanhavre** [comment]
Giving it a test spin ASAP!

**Comment by rostskadat** [comment]
> Giving it a test spin ASAP!

@yorikvanhavre  You can find samples at https://github.com/rostskadat/FreeCAD-SH3D/tree/main/Resources/sample if that helps

**Comment by yorikvanhavre** [comment]
This is pretty cool! Amazing results @rostskadat !

image

image

The import is specially well done, it uses BIM objects and it keeps all the SH3D data in properties. Great job!!

Are you planning to support the furniture? That might be cool! IIRC furniture in SH3D is stored as obj or any other common mesh format... It could be simple to support (in another future PR of course)

The project object might not be necessary? In fact in the BIM WB it is not used anymore.. Since its only usefulness was for IFC, it only makes sense when we are dealing with IFC files. For me you could remove it (unless there is some usefulness I don't know of course). No need to modify this PR I guess, but might be worth thinking for a next one :)

Overall, this PR is good for merge, but I'm not able to rebase it. In the 50 commits of this PR, it seems to include several commits from other branches that have already been merged, and git is apparently unable to sort things out...

Are you keen to keep all your commits? Otherwise since the work done in this PR is already scattered all over, I would simply take the diff from this PR and reapply it entirely to a new branch (which you can then force-push onto this branch)

Something like this:

```bash
# download a diff from this PR: Just add '.diff' to the URL:  https://patch-diff.githubusercontent.com/raw/FreeCAD/FreeCAD/pull/17165.diff
# make sure we start from a clean, up-to-date base (replace "origin" by the remote name of the FreeCAD repo)
git checkout main
git pull origin main
# create a new, clean branch
git checkout -b new-sh3d
# apply the diff
patch -p1 < ~/Downloads/17165.diff
# check that everything is OK
git status
git diff
# create a new commit
git add .
git commit -m "Improved SweetHome 3D importer"
# force-push to this branch (replace "github" by the remote name of your github repo)
git push -f github main_merge_conflict_17165
```

**Review by yorikvanhavre — APPROVED** [review]

**Comment by rostskadat** [comment]
> Overall, this PR is good for merge, but I'm not able to rebase it. In the 50 commits of this PR, it seems to include several commits from other branches that have already been merged, and git is apparently unable to sort things out...

@yorikvanhavre  thanks for your kind comments. Allow me to answer in order:

1. You are in fact able to import furniture. It is just not imported by default (maybe it *should* be imported by default) . In order to turn on the import furniture. You have to go to Edit >> Preferences >> Import-Export >> SH3D and use the settings there.
2. There are no problem in removing the Project object. It is not used. Once merged, I can create a new PR
3. I did merge / resolve conflicts while waiting for the merge to proceed. If it helps, all commit can be squashed. However I'm not sure whether you are requesting me to do that, or if it is something that the person merging the request would do?

Thanks for your feedback

**Comment by yorikvanhavre** [comment]
Okay I was able to squash! So no problem, forget about my proposed method above ;)
Thanks for the awesome work!!!

**Pull request merged**
